### PR TITLE
Implement FlightSqlProducer - PreparedStatement query

### PR DESCRIPTION
**Pull request by jcralmeida**
This Pull request changes the FlightProducer to FlightSqlProducer and implements that ability of query data via method from both producer

### CODE REVIEW

File: services/arrow-flight/src/main/java/com/dremio/service/flight/DremioFlightSqlProducer.java [hunk 1]
```diff
@@ -78,15 +107,29 @@ public void createPreparedStatement(
...
+    streamListener.onCompleted();
   }
 
   @Override
   public void closePreparedStatement(
     ActionClosePreparedStatementRequest actionClosePreparedStatementRequest,
     CallContext callContext,
     StreamListener<Result> listener) {
-    throw CallStatus.UNIMPLEMENTED.withDescription("closePreparedStatement not supported.").toRuntimeException();
+    // No need to close PreparedStatement
```
Comment (kylep-dremio) [inline services/arrow-flight/src/main/java/com/dremio/service/flight/DremioFlightSqlProducer.java:132]: ? Shouldn't you remove it from the cache?

File: services/arrow-flight/src/main/java/com/dremio/service/flight/DremioFlightSqlProducer.java [hunk 2]
```diff
@@ -122,7 +185,19 @@ public void getStreamPreparedStatement(
     CommandPreparedStatementQuery commandPreparedStatementQuery,
     CallContext callContext, Ticket ticket,
     ServerStreamListener serverStreamListener) {
-    throw CallStatus.UNIMPLEMENTED.withDescription("PreparedStatement not supported.").toRuntimeException();
+    final CallHeaders headers = retrieveHeadersFromCallContext(callContext);
+    final UserSession session = sessionsManager.getUserSession(callContext.peerIdentity(), headers);
+
+    try {
+      UserProtos.PreparedStatementHandle preparedStatementHandle =
+        UserProtos.PreparedStatementHandle.parseFrom(commandPreparedStatementQuery.getPreparedStatementHandle());
```
Comment (kylep-dremio) [inline services/arrow-flight/src/main/java/com/dremio/service/flight/DremioFlightSqlProducer.java:193]: Do you not need to validate that the prepared statement handle is in the cache?

File: services/arrow-flight/src/test/java/com/dremio/service/flight/TestFlightSqlServerWithBasicAuth.java [hunk 3]
```diff
@@ -21,14 +21,12 @@
 import org.apache.arrow.flight.FlightInfo;
 import org.apache.arrow.flight.sql.FlightSqlClient;
 import org.junit.BeforeClass;
-import org.junit.Ignore;
```
Comment (kylep-dremio) [inline services/arrow-flight/src/test/java/com/dremio/service/flight/TestFlightSqlServerWithBasicAuth.java:24]: Why are these changes made as part of this PR?

File: services/arrow-flight/src/main/java/com/dremio/service/flight/impl/FlightPreparedStatement.java [hunk 4]
```diff
@@ -65,6 +67,22 @@ public FlightInfo getFlightInfo(Location location) {
...
+    final Schema schema = buildSchema(createPreparedStatementResp.getPreparedStatement().getArrowSchema());
+
+    return ActionCreatePreparedStatementResult.newBuilder()
+        .setDatasetSchema(ByteString.copyFrom(schema.toByteArray()))
+        .setParameterSchema(ByteString.EMPTY)
+        .setPreparedStatementHandle(getServerHandle().toByteString())
+        .build();
+  }
+
+  public UserProtos.PreparedStatementHandle getServerHandle() {
```
Comment (kylep-dremio) [inline services/arrow-flight/src/main/java/com/dremio/service/flight/impl/FlightPreparedStatement.java:81]: Why public?

File: services/arrow-flight/src/main/java/com/dremio/service/flight/impl/FlightPreparedStatement.java [hunk 5]
```diff
@@ -65,6 +67,22 @@ public FlightInfo getFlightInfo(Location location) {
     return new FlightInfo(schema, flightDescriptor, ImmutableList.of(flightEndpoint), -1, -1);
   }
 
+  public ActionCreatePreparedStatementResult createAction() {
+    final UserProtos.CreatePreparedStatementArrowResp createPreparedStatementResp = responseHandler.get();
+    final Schema schema = buildSchema(createPreparedStatementResp.getPreparedStatement().getArrowSchema());
+
+    return ActionCreatePreparedStatementResult.newBuilder()
+        .setDatasetSchema(ByteString.copyFrom(schema.toByteArray()))
+        .setParameterSchema(ByteString.EMPTY)
```
Comment (kylep-dremio) [inline services/arrow-flight/src/main/java/com/dremio/service/flight/impl/FlightPreparedStatement.java:76]: Is this true?

File: services/arrow-flight/src/main/java/com/dremio/service/flight/DremioFlightSqlProducer.java (hunk 1 above)
Comment (rafael-telles) [reply services/arrow-flight/src/main/java/com/dremio/service/flight/DremioFlightSqlProducer.java]: You are right! Fixed

File: services/arrow-flight/src/main/java/com/dremio/service/flight/DremioFlightSqlProducer.java (hunk 2 above)
Comment (rafael-telles) [reply services/arrow-flight/src/main/java/com/dremio/service/flight/DremioFlightSqlProducer.java:197]: Fixed

File: services/arrow-flight/src/test/java/com/dremio/service/flight/TestFlightSqlServerWithBasicAuth.java (hunk 3 above)
Comment (rafael-telles) [reply services/arrow-flight/src/test/java/com/dremio/service/flight/TestFlightSqlServerWithBasicAuth.java:24]: In the previous PR we just refactored the producer without implementing any features, so these ignored test would fail when trying to make queries

File: services/arrow-flight/src/main/java/com/dremio/service/flight/impl/FlightPreparedStatement.java (hunk 4 above)
Comment (rafael-telles) [reply services/arrow-flight/src/main/java/com/dremio/service/flight/impl/FlightPreparedStatement.java:81]: Because it is part of "com.dremio.service.flight.impl" but used in "com.dremio.service.flight"

File: services/arrow-flight/src/main/java/com/dremio/service/flight/DremioFlightProducer.java [hunk 6]
```diff
@@ -76,6 +92,11 @@ public DremioFlightProducer(Location location, DremioFlightSessionsManager sessi
     this.allocator = allocator;
 
     flightWorkManager = new FlightWorkManager(workerProvider, optionManagerProvider, runQueryResponseHandlerFactory);
+
```
Comment (kylep-dremio) [inline services/arrow-flight/src/main/java/com/dremio/service/flight/DremioFlightProducer.java:95]: Nit: extra newline

File: services/arrow-flight/src/main/java/com/dremio/service/flight/DremioFlightProducer.java [hunk 7]
```diff
@@ -146,18 +217,39 @@ public void listActions(CallContext callContext, StreamListener<ActionType> stre
...
-    throw CallStatus.UNIMPLEMENTED.withDescription("closePreparedStatement not supported.").toRuntimeException();
+    UserProtos.PreparedStatementHandle preparedStatementHandle;
+    try {
+      preparedStatementHandle =
+        UserProtos.PreparedStatementHandle.parseFrom(actionClosePreparedStatementRequest.getPreparedStatementHandle());
+    } catch (InvalidProtocolBufferException e) {
+      throw CallStatus.INVALID_ARGUMENT.withDescription("Invalid PreparedStatementHandle").toRuntimeException();
+    }
+
+    flightPreparedStatementCache.invalidate(preparedStatementHandle);
```
Comment (kylep-dremio) [inline services/arrow-flight/src/main/java/com/dremio/service/flight/DremioFlightProducer.java:252]: Suggest you just put this in the try block.

File: services/arrow-flight/src/main/java/com/dremio/service/flight/DremioFlightProducer.java [hunk 8]
```diff
@@ -349,6 +433,21 @@ private CallHeaders retrieveHeadersFromCallContext(CallContext callContext) {
     return callContext.getMiddleware(FlightConstants.HEADER_KEY).headers();
   }
 
+  private UserSession getUserSessionFromCallContext(CallContext callContext) {
+    final CallHeaders headers = retrieveHeadersFromCallContext(callContext);
+    return sessionsManager.getUserSession(callContext.peerIdentity(), headers);
+  }
+
+  private <T extends Message> FlightInfo getFlightInfoForFlightSqlCommands(
+    T commandPreparedStatementQuery, FlightDescriptor flightDescriptor, FlightPreparedStatement preparedStatement) {
```
Comment (kylep-dremio) [inline services/arrow-flight/src/main/java/com/dremio/service/flight/DremioFlightProducer.java:442]: Is the variable name correct here?

File: services/arrow-flight/src/test/java/com/dremio/service/flight/TestFlightSqlServerWithBasicAuth.java [hunk 9]
```diff
@@ -28,7 +28,6 @@
 /**
  * Test FlightServer with basic authentication using FlightSql producer.
  */
-@Ignore
 public class TestFlightSqlServerWithBasicAuth extends AbstractTestFlightServer {
```
Comment (kylep-dremio) [inline services/arrow-flight/src/test/java/com/dremio/service/flight/TestFlightSqlServerWithBasicAuth.java:31]: Why was this change done as part of this PR, seems unrelated?

File: services/arrow-flight/src/main/java/com/dremio/service/flight/DremioFlightProducer.java (hunk 6 above)
Comment (rafael-telles) [reply services/arrow-flight/src/main/java/com/dremio/service/flight/DremioFlightProducer.java]: Fixed

File: services/arrow-flight/src/main/java/com/dremio/service/flight/DremioFlightProducer.java (hunk 7 above)
Comment (rafael-telles) [reply services/arrow-flight/src/main/java/com/dremio/service/flight/DremioFlightProducer.java]: Fixed

File: services/arrow-flight/src/main/java/com/dremio/service/flight/DremioFlightProducer.java (hunk 8 above)
Comment (rafael-telles) [reply services/arrow-flight/src/main/java/com/dremio/service/flight/DremioFlightProducer.java]: No.. sorry

Fixed

File: services/arrow-flight/src/test/java/com/dremio/service/flight/TestFlightSqlServerWithBasicAuth.java (hunk 9 above)
Comment (rafael-telles) [reply services/arrow-flight/src/test/java/com/dremio/service/flight/TestFlightSqlServerWithBasicAuth.java:31]: That's because in the previous PR there was no FlightSql functionality yet, so this tests would fail

File: services/arrow-flight/src/main/java/com/dremio/service/flight/impl/FlightPreparedStatement.java (hunk 5 above)
Comment (rafael-telles) [reply services/arrow-flight/src/main/java/com/dremio/service/flight/impl/FlightPreparedStatement.java]: Yes, FlightPreparedStatement doesn't support parameters

File: services/arrow-flight/src/main/java/com/dremio/service/flight/DremioFlightProducer.java [hunk 10]
```diff
@@ -245,11 +245,11 @@ public void closePreparedStatement(
     try {
       preparedStatementHandle =
```
Comment (kylep-dremio) [inline services/arrow-flight/src/main/java/com/dremio/service/flight/DremioFlightProducer.java:246]: No need to declare this variable outside the try block now.

Comment (rafael-telles) [reply services/arrow-flight/src/main/java/com/dremio/service/flight/DremioFlightProducer.java]: Fixed

File: services/arrow-flight/src/main/java/com/dremio/service/flight/DremioFlightProducer.java [hunk 11]
```diff
@@ -85,20 +105,46 @@ public void getStream(CallContext callContext, Ticket ticket, ServerStreamListen
...
+      preparedStatementHandle =
+        UserProtos.PreparedStatementHandle.parseFrom(commandPreparedStatementQuery.getPreparedStatementHandle());
+    } catch (InvalidProtocolBufferException e) {
+      throw CallStatus.INVALID_ARGUMENT.withDescription("Invalid PreparedStatementHandle").toRuntimeException();
+    }
+
+    // Check if given PreparedStatement is cached
+    FlightPreparedStatement preparedStatement = flightPreparedStatementCache.getIfPresent(preparedStatementHandle);
+    if (preparedStatement == null) {
+      throw CallStatus.NOT_FOUND.withDescription("PreparedStatement not found.").toRuntimeException();
```
Comment (jduo) [inline services/arrow-flight/src/main/java/com/dremio/service/flight/DremioFlightProducer.java:144]: Should this get written to serverStreamListener, or does the caller catch this and do it?

File: services/arrow-flight/src/main/java/com/dremio/service/flight/impl/FlightPreparedStatement.java [hunk 12]
```diff
@@ -15,6 +15,8 @@
  */
 package com.dremio.service.flight.impl;
 
+import static org.apache.arrow.flight.sql.impl.FlightSql.*;
```
Comment (jduo) [inline services/arrow-flight/src/main/java/com/dremio/service/flight/impl/FlightPreparedStatement.java:18]: Avoid star imports

File: services/arrow-flight/src/main/java/com/dremio/service/flight/DremioFlightProducer.java (hunk 11 above)
Comment (rafael-telles) [reply services/arrow-flight/src/main/java/com/dremio/service/flight/DremioFlightProducer.java:144]: The caller catches it and call `listener.error` with the thrown exception, this is done on `org.apache.arrow.flight.FlightService`

**Review by jduo — APPROVED** [review]
